# ĐỒ ÁN CUỐI KỲ LẬP TRÌNH CHO KHOA HỌC DỮ LIỆU

## Thông tin đồ án

+ Chủ đề: **Tìm hiểu về giá thuê chung cư ở thành phố Hồ Chí Minh**
+ Github: [p4ds-final](https://github.com/vanhung4499/p4ds-final)

### Thông tin nhóm

Nhóm chỉ có một thành viên duy nhất

| MSSV    | Họ và tên       |
|---------|-----------------|
| 1712222 | Nguyễn Văn Hùng |

## Môi trường làm việc

In [38]:
import sys
sys.executable

'/usr/local/anaconda3/envs/min_ds-env/bin/python'

Môi trường làm việc mà em dùng:
+ OS: macOS
+ IDE: Pycharm
+ Sử dụng mội trường `min_ds-evn` như các bài lab

## Import and constant

In [39]:
import numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sns

In [40]:
DATA_CSV = './data/hcm-apartment-rent-data.csv'

## 1. Thu thập dữ liệu

### 1.1. Chủ đề, mục đích và nguồn của dữ liệu?

+ Chủ đề em nhắm tới là về giá thuê chung cư ở thành phố Hồ Chí Minh
+ Do tìm trên kaggle chỉ có duy nhất một bộ dữ liệu về giá thuê chung cư của thành phố Hồ Chí Minh nhưng đã cũ và em cảm thấy nó ít không đủ các cột nên em đã tự crawl dữ liệu từ [nhatot.vn](https://www.nhatot.com/thue-can-ho-chung-cu-tp-ho-chi-minh), trang môi giới xếp tốp đầu trong lĩnh vực bất động sản trên [Google](https://www.google.com/search?q=thu%C3%AA+chung+c%C6%B0+h%E1%BB%93+ch%C3%AD+minh&oq=thu%C3%AA+chung+c%C6%B0&aqs=chrome.1.69i59l3j0i512l2j69i60l3.4604j0j7&sourceid=chrome&ie=UTF-8)
+ Mục đích của em khá đơn giản là để xem xét giá cả chung cư cho thuê ở TPHCM, vì trong thời gian tới em cũng sẽ tìm chỗ ở mới (làm trước sau dùng lại)

### 1.2. Người ta có cho phép sử dụng dữ liệu như thế này hay không?

+ Việc crawl dữ liệu từ nhatot có thể là không hợp pháp, trang web cũng sử dụng công cụ chống ddos, nhưng mà để phục vụ cho đồ án thì chắc không sao cả!

### 1.3. Thu thập dữ liệu này như thế nào?

+ Dữ liệu được em crawl từ [nhatot.vn](https://www.nhatot.com/thue-can-ho-chung-cu-tp-ho-chi-minh), là trang môi giới về bất động sản lớn ở Việt Nam, web có rất nhiều tin cho thuê nhà
+ Em crawl dữ liệu toàn bộ 1000 trang thu về 20000 record
+ Chi tiết về cách crawl được mô tả ở file notebook `Crawl-data.ipynb`

## 2. Khám phá dữ liệu

### 2.1. Đọc dữ liệu

In [43]:
apartment_df = pd.read_csv(DATA_CSV)
apartment_df.head()

ad_id    list_id      list_time  \
0  139905225  101449743  1669427357000   
1  140064167  101580825  1669783016000   
2  139905234  101449753  1669427371000   
3  140064175  101580838  1669783034000   
4  140064185  101580839  1669783034000   

                                             subject  \
0       Căn hộ CMT8 Q10, Cửa sổ, Máy giặt riêng, Gác   
1     Căn Hộ 1PN _ Kcx Tân Thuận chợ UFM sát Quận 4🌈   
2            KHAI TRƯƠNG CĂN HỘ BANCOL FULL NỘI THẤT   
3  CĂN HỘ  ĐẦY ĐỦ NỘI THẤT GẦN ĐH KINH TẾ 👉 CHỈ 3TR7   
4   Căn hộ ban công lớn cực Chill_Ngay Điện Biên Phủ   

                                                body       price  \
0  Căn hộ Cách Mạng Tháng 8 Q10, Cửa sổ, Máy giặt...   9000000.0   
1  ⭐️Sắp hết tháng rồi khách đã tìm được phòng ch...   5700000.0   
2  Cho thuê căn hộ\n_Vị trí : đường Thạch Lam p.p...   4500000.0   
3  💥💥CĂN HỘ  ĐẦY ĐỦ NỘI THẤT GẦN ĐH KINH TẾ\n🍀 Vị...   3700000.0   
4  📣📣 HỆ THỐNG CĂN HỘ STUDIO, 1PN CÒN TRỐNG, FULL...  11000000.0   

      price_string  size  rooms  toilets          area_name     region_name  \
0    9 triệu/tháng  30.0      1      1.0            Quận 10  Tp Hồ Chí Minh   
1  5,7 triệu/tháng  37.0      1      1.0             Quận 7  Tp Hồ Chí Minh   
2  4,5 triệu/tháng  35.0      1      1.0       Quận Tân Phú  Tp Hồ Chí Minh   
3  3,7 triệu/tháng  30.0      1      1.0  Quận Ngũ Hành Sơn         Đà Nẵng   
4   11 triệu/tháng  45.0      1      1.0    Quận Bình Thạnh  Tp Hồ Chí Minh   

           ward_name  deposit  zero_deposit  longitude  latitude  owner  \
0          Phường 13        0         False  106.67245  10.78176  False   
1  Phường Bình Thuận        0         False  106.72563  10.74310  False   
2   Phường Phú Thạnh        0         False  106.62711  10.77815  False   
3     Phường Khuê Mỹ        0         False  108.25112  16.03646  False   
4          Phường 17        0         False  106.70816  10.79913  False   

   company_ad  apartment_type  
0        True  Căn hộ dịch vụ  
1        True  Căn hộ dịch vụ  
2        True  Căn hộ dịch vụ  
3        True  Căn hộ dịch vụ  
4        True        Chung cư

### 2.2. Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [45]:
apartment_df.shape

(19996, 20)

Dữ liệu gồm 19996 dòng và 20 cột, đúng như những gì đã thu thập trước đó

### 2.3. Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

+ Mỗi dòng là thông tin cho thuê của một căn hộ như giá thuê, diện tích, số phòng, số nhà vệ sinh, địa chỉ, ...
+ Dựa theo thông tin trên web lúc crawl thì sẽ không có các dòng có ý nghĩa khác nhau

### 2.4. Mỗi cột có ý nghĩa gì?

Dưới đây là phần mô tả về các cột:

| Column         | Description                                |
|----------------|--------------------------------------------|
| ad_id          | id của tin (để phân biệt)                  |
| list_id        | id trong danh sách ads (dùng để phân biệt) |
| list_time      | Thời gian đăng tin                         |
| price          | Giá thuê (VND)                             |
| price_string   | Giá thuê (triệu / tháng) cho dễ nhìn       |
| size           | Diện tích (m2)                             |
| rooms          | Số phòng                                   |
| toilets        | Số nhà vệ sinh                             |
| body           | Mô tả                                      |
| street_name    | Đường                                      |
| ward_name      | Phường                                     |
| area_name      | Quận huyện                                 |
| region_name    | Thành phố, tỉnh                            |
| zero_deposit   | Có cần tiền cọc không?                     |
| deposit        | Tiền cọc                                   |
| longitude      | Kinh độ                                    |
| latitude       | Vĩ độ                                      |
| owner          | Có phải chủ nhà không?                     |
| company_ad     | Tin của các nhân hay công ty               |
| apartment_type | Kiểu căn hộ / chung cư                     |

### 2.4. Kiểm tra dữ liệu có bị trùng lặp không?

In [48]:
apartment_df.duplicated().sum()

0

Dữ liệu thu thập được không bị trùng lặp, dựa vào việc crawl trước đó nên có thể yên tâm

### 2.5. Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [49]:
apartment_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19996 entries, 0 to 19995
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ad_id           19996 non-null  int64  
 1   list_id         19996 non-null  int64  
 2   list_time       19996 non-null  int64  
 3   subject         19996 non-null  object 
 4   body            19996 non-null  object 
 5   price           19979 non-null  float64
 6   price_string    19979 non-null  object 
 7   size            19995 non-null  float64
 8   rooms           19996 non-null  int64  
 9   toilets         19989 non-null  float64
 10  area_name       19990 non-null  object 
 11  region_name     19996 non-null  object 
 12  ward_name       19989 non-null  object 
 13  deposit         19996 non-null  int64  
 14  zero_deposit    19996 non-null  bool   
 15  longitude       19987 non-null  float64
 16  latitude        19987 non-null  float64
 17  owner           19996 non-null 

In [8]:
apartment_df.describe().T

count          mean           std           min  \
ad_id               20000.0  1.367747e+08  7.252702e+06  1.156086e+08   
list_id             20000.0  9.891263e+07  5.859449e+06  8.187637e+07   
price               20000.0  1.295000e+07  1.098415e+07  4.200000e+06   
size                20000.0  6.337500e+01  2.946380e+01  2.500000e+01   
rooms               20000.0  1.750000e+00  6.982295e-01  1.000000e+00   
toilets             20000.0  1.600000e+00  5.831098e-01  1.000000e+00   
escrow_can_deposit  20000.0  2.000000e+00  0.000000e+00  2.000000e+00   
longitude           20000.0  1.066956e+02  4.679639e-02  1.066037e+02   
latitude            20000.0  1.077714e+01  4.450595e-02  1.070316e+01   

                             25%           50%           75%           max  
ad_id               1.395121e+08  1.400724e+08  1.400809e+08  1.400811e+08  
list_id             1.011230e+08  1.015880e+08  1.015949e+08  1.015950e+08  
price               5.725000e+06  8.250000e+06  1.625000e+07  4.900000e+07  
size                3.375000e+01  6.900000e+01  7.825000e+01  1.280000e+02  
rooms               1.000000e+00  2.000000e+00  2.000000e+00  3.000000e+00  
toilets             1.000000e+00  2.000000e+00  2.000000e+00  3.000000e+00  
escrow_can_deposit  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00  
longitude           1.066633e+02  1.067123e+02  1.067349e+02  1.067553e+02  
latitude            1.074063e+01  1.078563e+01  1.080627e+01  1.086830e+01

## 2. Xử lý dữ liệu

##